mam/v1/mam.go

```diff
@@ -0,0 +1,298 @@
+// Package mam provides functions for creating Masked Authentication Messaging messages
+package mam
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	. "github.com/iotaledger/iota.go/consts"
+	"github.com/iotaledger/iota.go/curl"
+	. "github.com/iotaledger/iota.go/curl/hamming"
+	. "github.com/iotaledger/iota.go/merkle"
+	"github.com/iotaledger/iota.go/signing/legacy"
+	"github.com/iotaledger/iota.go/signing/utils"
+	. "github.com/iotaledger/iota.go/trinary"
+)
+
+// MamInitEncryption initializes the encryption/decryption state for a MAM session
+//
+//	sideKey is the encryption/decryption key
+//	merkleRoot is the merkle root
+//	spongeFunc is the spongeFunction instance used for encryption/decryption
+func MamInitEncryption(sideKey Trits, merkleRoot Trits, spongeFunc sponge.SpongeFunction) {
+	spongeFunc.Absorb(sideKey)
+	spongeFunc.Absorb(merkleRoot[:HashTrinarySize])
+}
+
+// PayloadMinLength computes the minimum length of a payload
+//
+//	messageLength is the length of the message
+//	merkleTreeLength is the length of the merkle tree
+//	index is the leaf index
+//	security is the security level used to generate addresses
+//
+// Returns the minimum length of the payload
+func PayloadMinLength(messageLength, merkleTreeLength, index uint64, security SecurityLevel) uint64 {
+	siblingNumber := MerkleDepth(merkleTreeLength/HashTrinarySize) - 1
+	return EncodedLength(int64(index)) + EncodedLength(int64(messageLength)) +
+		HashTrinarySize + messageLength + HashTrytesSize +
+		uint64(security)*ISSKeyLength + EncodedLength(int64(siblingNumber)) +
+		siblingNumber*HashTrinarySize
+}
+
+// MamCreate creates a signed, encrypted payload from a message.
+//
+//	payloadLength is the length of the payload
+//	message is the message to encrypt
+//	sideKey is the encryption key
+//	merkleTree is the merkle tree
+//	merkleTreeLength is the length of the merkle tree
+//	leafCount is the number of leaves of the merkle tree
+//	index is the leaf index
+//	nextRoot is the merkle root of the next MAM payload
+//	start is the offset used to generate addresses
+//	seed is the seed used to generate addresses - Not sent over the network
+//	security is the security level used to generate addresses
+//
+// Returns:
+//  payload is the payload of the encrypted message
+//  payloadMinLength is the length of the payload needed to encrypt the message
+//	err is the error message
+func MamCreate(payloadLength uint64,
+	message Trytes, sideKey Trytes,
+	merkleTree Trits, merkleTreeLength uint64,
+	leafCount uint64, index uint64,
+	nextRoot Trits, start uint64,
+	seed Trytes, security SecurityLevel) (payload Trits, payloadMinLength uint64, err error) {
```

Comment:
Declare the return variables inside the function instead in the function signature.